src/gradualizer.erl

```diff
@@ -56,21 +56,47 @@ type_check_file(File) ->
 %% @doc Type check a source or beam file
 -spec type_check_file(file:filename(), options()) -> ok | nok | [{file:filename(), any()}].
 type_check_file(File, Opts) ->
-    ParsedFile =
-        case filename:extension(File) of
-            ".erl" ->
-                Includes = proplists:get_all_values(i, Opts),
-                gradualizer_file_utils:get_forms_from_erl(File, Includes);
-            ".beam" ->
-                gradualizer_file_utils:get_forms_from_beam(File);
-            Ext ->
-                throw({unknown_file_extension, Ext})
-        end,
-    case ParsedFile of
-        {ok, Forms} ->
+    case filename:extension(File) of
+        ".erl" ->
+            Includes = proplists:get_all_values(i, Opts),
+            case gradualizer_file_utils:get_forms_from_erl(File, Includes) of
+                {ok, Forms} ->
+                    lint_and_check_forms(Forms, File, Opts);
+                Error ->
+                    throw(Error) % TODO: Don't use throw. These are user errors.
+            end;
+        ".beam" ->
+            case gradualizer_file_utils:get_forms_from_beam(File) of
+                {ok, Forms} ->
+                    type_check_forms(File, Forms, Opts);
+                Error ->
+                    throw(Error) % TODO: Don't use throw. These are user errors.
+            end;
+        Ext ->
+            throw({unknown_file_extension, Ext}) %% TODO: Don't use throw here.
```

Comment:
Can you create a ticket and add the ticket number to the TODOs so that we keep track of that we need to remove the throws in the future?

Reply:
Too much effort to describe it, including context and consequences, in an issue. I deleted the TODOs instead.